kvbc/src/kvbc_adapter/app_state_adapter.cpp

```diff
@@ -0,0 +1,221 @@
+        histograms_.get_block_duration->recordAtomic(jobDuration);
+        return result;
+      },
+      std::forward<decltype(blockId)>(blockId),
+      std::forward<decltype(outBlock)>(outBlock),
+      std::forward<decltype(outBlockMaxSize)>(outBlockMaxSize),
+      std::forward<decltype(outBlockActualSize)>(outBlockActualSize));
+
+  if ((++callCounter % snapshotThresh) == 0) {
+    auto &registrar = concord::diagnostics::RegistrarSingleton::getInstance();
+    registrar.perf.snapshot("iappstate");
+    LOG_INFO(logger_, registrar.perf.toString(registrar.perf.get("iappstate")));
+  }
+  return future;
+}
+bool AppStateAdapter::getPrevDigestFromBlock(uint64_t blockId,
+                                             bftEngine::bcst::StateTransferDigest *outPrevBlockDigest) const {
+  ConcordAssert(blockId > 0);
+  const auto parent_digest = kvbc_->parentDigest(blockId);
+
+  if (!parent_digest.has_value()) {
+    LOG_WARN(logger_, "parent digest not found," << KVLOG(blockId));
+    return false;
+  }
+  static_assert(parent_digest->size() == DIGEST_SIZE);
+  static_assert(sizeof(bftEngine::bcst::StateTransferDigest) == DIGEST_SIZE);
+  std::memcpy(outPrevBlockDigest, parent_digest->data(), DIGEST_SIZE);
+  return true;
+}
+void AppStateAdapter::getPrevDigestFromBlock(const char *blockData,
+                                             const uint32_t blockSize,
+                                             bftEngine::bcst::StateTransferDigest *outPrevBlockDigest) const {
+  ConcordAssertGT(blockSize, 0);
+  auto view = std::string_view{blockData, blockSize};
+  const auto rawBlock = categorization::RawBlock::deserialize(view);
+
+  static_assert(rawBlock.data.parent_digest.size() == DIGEST_SIZE);
+  static_assert(sizeof(bftEngine::bcst::StateTransferDigest) == DIGEST_SIZE);
+  std::memcpy(outPrevBlockDigest, rawBlock.data.parent_digest.data(), DIGEST_SIZE);
+}
+bool AppStateAdapter::putBlock(const uint64_t blockId,
+                               const char *blockData,
+                               const uint32_t blockSize,
+                               bool lastBlock) {
+  auto view = std::string_view{blockData, blockSize};
+  const auto rawBlock = categorization::RawBlock::deserialize(view);
+  if (kvbc_->hasBlock(blockId)) {
+    const auto existingRawBlock = kvbc_->getRawBlock(blockId);
+    if (rawBlock != existingRawBlock) {
+      LOG_ERROR(logger_,
+                "found existing (and different) block ID[" << blockId << "] when receiving from state transfer");
+
+      // TODO consider assert?
+      kvbc_->deleteBlock(blockId);
+      throw std::runtime_error(
+          __PRETTY_FUNCTION__ +
+          std::string("found existing (and different) block when receiving state transfer, block ID: ") +
+          std::to_string(blockId));
+    }
+  } else {
+    kvbc_->addRawBlock(rawBlock, blockId, lastBlock);
+  }
+  return true;
+}
+std::future<bool> AppStateAdapter::putBlockAsync(uint64_t blockId,
```

Comment:
seems that all this code can be in a common wrapper.

Reply:
Thanks @cloudnoize, I have moved the putBlockAsync and getBlockAsync in the common wrapper (replica_adapter)